system/Encryption/Encryption.php

```diff
@@ -0,0 +1,298 @@
+		{
+			throw EncryptionException::forDriverNotAvailable($params['driver']);
+		}
+
+		// Derive a secret key for the encrypter
+		if (isset($params['key']))
+		{
+			$hmacKey          = strcmp(phpversion(), '7.1.2') >= 0 ? \hash_hkdf($this->digest, $params['key']) : self::hkdf($params['key'], $this->digest);
+			$params['secret'] = bin2hex($hmacKey);
+		}
+
+		$handlerName     = 'CodeIgniter\\Encryption\\Handlers\\' . $this->driver . 'Handler';
+		$this->encrypter = new $handlerName($params);
+		return $this->encrypter;
+	}
+
+	/**
+	 * Determine proper parameters
+	 *
+	 * @param array|object $params
+	 *
+	 * @return array|null
+	 */
+	protected function properParams($params = null)
+	{
+		// use existing config if no parameters given
+		if (empty($params))
+		{
+			$params = $this->config;
+		}
+
+		// treat the paramater as a Config object?
+		if (is_object($params))
+		{
+			$params = (array) $params;
+		}
+
+		// override base config with passed parameters
+		$params = array_merge($this->config, $params);
+		// make sure we only have expected parameters
+		$params = array_intersect_key($params, $this->default);
+
+		// and remember what we are up to
+		$this->config = $params;
+
+		// make the parameters conveniently accessible
+		foreach ($params as $pkey => $value)
+		{
+			$this->$pkey = $value;
+		}
+
+		return $params;
+	}
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Create a random key
+	 *
+	 * @param  integer $length Output length
+	 * @return string
+	 */
+	public static function createKey($length = 32)
+	{
+		return random_bytes($length);
+	}
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * __get() magic, providing readonly access to some of our protected properties
+	 *
+	 * @param  string $key Property name
+	 * @return mixed
+	 */
+	public function __get($key)
+	{
+		if (in_array($key, ['config', 'key', 'driver', 'drivers', 'default'], true))
+		{
+			return $this->{$key};
+		}
+
+		return null;
+	}
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * Byte-safe strlen()
+	 *
+	 * @param  string $str
+	 * @return integer
+	 */
+	protected static function strlen($str)
+	{
+		return mb_strlen($str, '8bit');
+	}
+
+	// --------------------------------------------------------------------
+
+	/**
+	 * HKDF legacy implementation, from CodeIgniter3.
+	 *
+	 * Fallback if PHP version < 7.1.2
+	 *
+	 * @link https://tools.ietf.org/rfc/rfc5869.txt
+	 *
+	 * @param string  $key    Input key
+	 * @param string  $digest A SHA-2 hashing algorithm
+	 * @param string  $salt   Optional salt
+	 * @param integer $length Output length (defaults to the selected digest size)
+	 * @param string  $info   Optional context/application-specific info
+	 *
+	 * @return string    A pseudo-random key
+	 */
+	public static function hkdf($key, $digest = 'sha512', $salt = null, $length = 64, $info = '')
```

Comment:
Since our minimum is PHP 7.2 now, can we kill this?

Reply:
done